### PR TITLE
did stuff with heartbeats and errors

### PR DESCRIPTION
**Pull request by supsm**
You should know from discord

### CODE REVIEW

File: discordpp/bot.hh [hunk 1]
```diff
@@ -197,12 +195,11 @@ class Bot : public virtual BotStruct {
                  // heartbeat that was received
             needACK_ = false;
             if(showHeartbeats) {
-                std::cout << "Heartbeat Successful." << std::endl;
+                log::logheartbeat([](std::ostream* log) { *log << "Heartbeat Sucessful.\n"; });
```
Comment (UndarkAido) [inline discordpp/bot.hh:198]: Spelling

File: discordpp/botStruct.hh [hunk 2]
```diff
@@ -27,9 +28,9 @@ class BotStruct {
                       sptr<const handleSent> callback) = 0;
 
     void run() {
-        for (const auto& module : needInit) {
+        for (auto module : needInit) {
```
Comment (UndarkAido) [inline discordpp/botStruct.hh:31]: Why?

File: discordpp/log.hh [hunk 3]
```diff
@@ -15,12 +15,22 @@ using handleLog = std::function<void(std::ostream *)>;
...
 inline std::ostream *out = &std::cout;
+inline std::ostream* err = &std::cerr;
+inline std::ostream* heartbeat = &std::cout;
 
 inline void log(level l, handleLog msg) {
     if (l >= filter) {
         msg(out);
     }
 }
+inline void logerr(handleLog msg)
```
Comment (UndarkAido) [inline discordpp/log.hh:26]: Just make everything use the standard `log`

File: discordpp/log.hh [hunk 4]
```diff
@@ -15,12 +15,22 @@ using handleLog = std::function<void(std::ostream *)>;
...
+{
+    msg(err);
+}
+inline void logheartbeat(handleLog msg)
```
Comment (UndarkAido) [inline discordpp/log.hh:30]: I specifically said not to touch the heartbeats?

**Review by UndarkAido — CHANGES_REQUESTED** [review]

File: discordpp/botStruct.hh (hunk 2 above)
Comment (supsm) [reply discordpp/botStruct.hh:31]: uh... I don't recall changing this

File: discordpp/log.hh (hunk 4 above)
Comment (supsm) [reply discordpp/log.hh:30]: ok, I'll remove it

File: discordpp/bot.hh [hunk 5]
```diff
@@ -96,7 +96,7 @@ class Bot : public virtual BotStruct {
         }
         needACK_++;
         if(showHeartbeats) {
-            std::cout << "Sending heartbeat..." << std::endl;
+            std::cout << "Sending heartbeat...\n";
```
Comment (UndarkAido) [inline discordpp/bot.hh:99]: Heartbeat should stay `endl`

File: discordpp/bot.hh [hunk 6]
```diff
@@ -119,8 +119,7 @@ class Bot : public virtual BotStruct {
             sequence_ = payload["s"].get<int>();
             if (handlers.find(payload["t"]) == handlers.end()) {
                 if (debugUnhandled) {
-                    std::cerr << "No handlers defined for " << payload["t"]
-                              << "\n";
+                    log::log(log::error, [payload](std::ostream* log) { *log << "No handlers defined for " << payload["t"] << "\n"; });
```
Comment (UndarkAido) [inline discordpp/bot.hh:122]: Oh wait this one has its own flag, it should be left alone.

File: discordpp/bot.hh [hunk 7]
```diff
@@ -130,17 +129,16 @@ class Bot : public virtual BotStruct {
             }
             break;
         case 1: // Heartbeat:          used for ping checking
-            std::cerr << "Discord Servers requested a heartbeat, which is not "
-                         "implemented.\n";
+            log::log(log::error, [](std::ostream* log) { *log << "Discord Servers requested a heartbeat, which is not implemented.\n"; });
```
Comment (UndarkAido) [inline discordpp/bot.hh:132]: Make this an `info`

File: discordpp/bot.hh [hunk 8]
```diff
@@ -130,17 +129,16 @@ class Bot : public virtual BotStruct {
...
             break;
         case 7: // Reconnect:          used to tell clients to reconnect to the
                 // gateway
-            std::cerr << "Discord Servers requested a reconnect.\n";
+            log::log(log::error, [](std::ostream* log) { *log << "Discord Servers requested a reconnect.\n"; });
```
Comment (UndarkAido) [inline discordpp/bot.hh:136]: `info`

File: discordpp/bot.hh [hunk 9]
```diff
@@ -130,17 +129,16 @@ class Bot : public virtual BotStruct {
...
             reconnect("The gateway is restarting");
             break;
         case 9: // Invalid Session:	used to notify client they have an
                 // invalid session id
-            std::cerr << "Discord Servers notified of an invalid session ID.\n";
+            log::log(log::error, [](std::ostream* log) { *log << "Discord Servers notified of an invalid session ID.\n"; });
```
Comment (UndarkAido) [inline discordpp/bot.hh:141]: `warning`

File: discordpp/bot.hh [hunk 10]
```diff
@@ -197,12 +195,11 @@ class Bot : public virtual BotStruct {
                  // heartbeat that was received
             needACK_ = false;
             if(showHeartbeats) {
-                std::cout << "Heartbeat Successful." << std::endl;
+                std::cout << "Heartbeat Sucessful.\n";
```
Comment (UndarkAido) [inline discordpp/bot.hh:198]: Should also stay `endl`

File: discordpp/bot.hh [hunk 11]
```diff
@@ -197,12 +195,11 @@ class Bot : public virtual BotStruct {
...
             }
             break;
         default:
-            std::cerr << "Unexpected opcode " << payload["op"] << "! Message:\n"
-                      << payload.dump(4) << '\n';
+            log::log(log::error, [payload](std::ostream* log) { *log << "Unexpected opcode " << payload["op"] << "! Message:\n" << payload.dump(4) << '\n'; });
```
Comment (UndarkAido) [inline discordpp/bot.hh:202]: `warning`

File: discordpp/botStruct.hh [hunk 12]
```diff
@@ -38,12 +39,12 @@ class BotStruct {
...
   protected:
     virtual void runctd() {
-        std::cerr << "Starting run loop" << '\n';
+        log::log(log::error, [](std::ostream* log) { *log << "Starting run loop" << '\n'; });
         work = std::make_unique<boost::asio::executor_work_guard<
             boost::asio::io_context::executor_type>>(
             boost::asio::make_work_guard(*aioc));
         aioc->run();
-        std::cerr << "Ending run loop" << '\n';
+        log::log(log::error, [](std::ostream* log) { *log << "Ending run loop" << '\n'; });
```
Comment (UndarkAido) [inline discordpp/botStruct.hh:47]: Hmm... These should either be left as-is. `'\n'` could be moved into the strings though.

**Review by UndarkAido — CHANGES_REQUESTED** [review]

**Pull request merged**